src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -184,103 +287,193 @@ public String getCredentialsId() {
+        if (endpoint != null) {
+            // we have a match
+            setServerUrl(endpoint.getServerUrl());
+            return;
+        }
+        LOGGER.log(Level.WARNING, "Call to legacy setBitbucketServerUrl({0}) method is configuring an url missing "
+                + "from the global configuration.", url);
+        setServerUrl(url);
     }
 
-    public String getRepository() {
-        return repository;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    @CheckForNull
+    public String getBitbucketServerUrl() {
+        String serverUrl = getServerUrl();
+        if (BitbucketEndpointConfiguration.get().findEndpoint(serverUrl) instanceof BitbucketCloudEndpoint) {
+            return null;
+        }
+        return serverUrl;
+    }
+
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    @CheckForNull
+    public String getCheckoutCredentialsId() {
+        for (SCMSourceTrait t : traits) {
+            if (t instanceof SSHCheckoutTrait) {
+                return StringUtils.defaultString(((SSHCheckoutTrait) t).getCredentialsId(), DescriptorImpl.ANONYMOUS);
+            }
+        }
+        return DescriptorImpl.SAME;
     }
 
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
     @DataBoundSetter
-    public void setAutoRegisterHook(boolean autoRegisterHook) {
-        this.autoRegisterHook = autoRegisterHook;
+    public void setCheckoutCredentialsId(String checkoutCredentialsId) {
+        for (Iterator<SCMSourceTrait> iterator = traits.iterator(); iterator.hasNext(); ) {
+            if (iterator.next() instanceof SSHCheckoutTrait) {
+                iterator.remove();
+            }
+        }
+        if (checkoutCredentialsId != null && !DescriptorImpl.SAME.equals(checkoutCredentialsId)) {
+            traits.add(new SSHCheckoutTrait(checkoutCredentialsId));
+        }
     }
 
-    public boolean isAutoRegisterHook() {
-        return autoRegisterHook;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    @NonNull
+    public String getIncludes() {
+        for (SCMSourceTrait trait : traits) {
+            if (trait instanceof WildcardSCMHeadFilterTrait) {
+                return ((WildcardSCMHeadFilterTrait) trait).getIncludes();
+            }
+        }
+        return "*";
     }
 
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
     @DataBoundSetter
-    public void setBitbucketServerUrl(String url) {
-        this.bitbucketServerUrl = Util.fixEmpty(url);
-        if (this.bitbucketServerUrl != null) {
-            // Remove a possible trailing slash
-            this.bitbucketServerUrl = this.bitbucketServerUrl.replaceAll("/$", "");
+    public void setIncludes(@NonNull String includes) {
+        for (int i = 0; i < traits.size(); i++) {
```

Comment:
This code is all legacy code. I'd delete it only it would break backwards binary compatibility. Much rather just leave the copy&paste partial duplication (because the type of list is different you wouldn't save as much moving to a utility class) and in any case there is test coverage of these methods